draft-ietf-tls-esni.md

```diff
@@ -723,38 +716,56 @@ error code.
+the second. Likewise, if the client did not offer ECH in the first ClientHello,
+then it MUST NOT not offer ECH in the second.
+
+#### Accepted ECH {#client-hrr-accepted-ech}
 
-1. TLS 1.3 {{!RFC8446}} ciphersuites in the ClientHello.cipher_suites list.
-1. The "key_share" and "supported_groups" extensions {{RFC8446}}. (These
-extensions may be copied from ClientHelloOuter into ClientHelloInner as
-described in {{real-ech}}.)
-1. Versions in the "supported_versions" extension, excluding TLS 1.2 and
-earlier. Note the ClientHelloOuter MAY include these older versions, while the
-ClientHelloInner MUST omit them.
+As described in {{sever-hrr}, the HelloRetryRequestInner handshake message is
+encrypted by the backend server and included as the payload of the
+"encrypted_client_hello" extension of the HelloRetryRequestOuter. The client
+decrypts the payload using the HPKE context it shares with the client-facing
+server. It begins by exporting a key and nonce for AEAD decryption:
 
-Future extensions that might change across first and second ClientHello messages
-in response to a HelloRetryRequest MUST have the same value.
+~~~
+    hrr_inner_key = context.Export("hrr inner key", Nk)
+    hrr_inner_nonce = context.Export("hrr inner nonce", Nn)
+~~~
 
-If the server sends a HelloRetryRequest in response to the ClientHello, the
-client sends a second updated ClientHello per the rules in {{RFC8446}}.
-However, at this point, the client does not know whether the server processed
-ClientHelloOuter or ClientHelloInner, and MUST regenerate both values to be
-acceptable. Note: if ClientHelloOuter and ClientHelloInner use different groups
-for their key shares or differ in some other way, then the HelloRetryRequest
-may actually be invalid for one or the other ClientHello, in which case a fresh
-ClientHello MUST be generated, ignoring the instructions in HelloRetryRequest.
-Otherwise, the usual rules for HelloRetryRequest processing apply.
-
-The client encodes the second ClientHelloInner as in {{encoding-inner}}, using
-the second ClientHelloOuter for any referenced extensions. It then encrypts
-the new EncodedClientHelloInner value as a second message with the previous
-HPKE context:
+Here `Nk` and `Nn` denote, respectively, the key and nonce length of the AEAD
+algorithm associated with `context`. Let `encrypted_hrr_inner` denote the
+extension payload. The client computes
+
+~~~
+    HelloRetryRequestInner = Open(hrr_inner_key,
+                                  hrr_inner_nonce,
+                                  "", // Empty AAD
+                                  encrypted_hrr_inner)
+~~~
+
+If decryption fails, then the client aborts with a "illegal_parameter" alert.
+
+The second ClientHelloOuter is identical to the first, except that the
+extensions field contains only the "encrypted_client_hello" extension, which
+contains the second ClientHelloInner. The client uses HelloRetryRequestInner to
```

Comment:
This seems like it's a problem in that it will have the wrong version, for instance.

Reply:
In this context the connection is being terminated by the backend server. Thus, the parameters of CHOuter2 don't need to change in response to the HRR. We thought it would be best if CHOuter1 and CHOuter2 had the same parameters. In fact, below we say that the HRROuter SHOULD NOT contain extensions other than "cookie" and "encrypted_client_hello".

Reply:
That's what I'm complaining about. It's not a valid TLS 1.3 CH because it doesn't have supported_versions.

Reply:
Ah crap. Good catch. Are there any other extensions you can think of that are required?